src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -175,10 +177,12 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
                     if (diagIdLength >= 0)
                     {
                         ReadOnlyMemory<char> idSlice = key.AsMemory().Slice(DiagnosticOptionPrefix.Length, diagIdLength);
+                        // PERF: this is similar to a double-checked locking pattern, and trying to fetch the ID first
+                        // lets us avoid an allocation if the id has already been added
```

Comment:
Why not use the `GetOrAdd` which takes a delegate? That would avoid the double lookup here and still allow for delay creation of the value.

Reply:
Wouldn't that requiring capturing the ReadOnlyMemory, creating a closure and allocating anyway?

Reply:
The `ReadOnlyMemory<char>` could be passed as the factory argument . That would allow this to target a `static` method which could be explicitly, or just implicitly, cached. 

---
In reply to: [322377924](https://github.com/dotnet/roslyn/pull/38592#discussion_r322377924) [](ancestors = 322377924)

Reply:
Ah, never mind, I misunderstood the signature. Let me try this.

Reply:
Hmm, no I'm still not seeing it, even with the third overload. If we use

```C#
public TValue GetOrAdd (TKey key, Func<TKey,TValue> valueFactory);
```

This is wrong, because the key will be the ROM from the file, keeping the file alive.

If we use

```C#
public TValue GetOrAdd<TArg> (TKey key, Func<TKey,TArg,TValue> valueFactory, TArg factoryArgument);
```

there's still no way to give a new value for the key, which needs to be the ROM for the newly allocated string.

Reply:
Interesting! And thanks for adding the PERF comment. 👍 

Consider expanding the comment to clarify the difference between lookup and keys:

> We use ReadOnlyMemory\<char> to allow allocation-free lookups in the dictionary, but the actual keys stored in the dictionary are trimmed to avoid holding GC references to larger strings than necessary. The GetOrAdd APIs do not allow the key to be manipulated between lookup and insertion, so we separate the operations here in code.